src/tokio_reactor.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    Config, Error, MemoryKeychain, Message, Moderator, Notification, Reactor, Request, Response,
+};
+
+use async_trait::async_trait;
+use tokio::sync::mpsc;
+
+use core::time::Duration;
+use std::thread;
+use std::time::SystemTime;
+
+/// Communication bridge with a consensus reactor.
+pub struct TokioReactor {
+    timeout: Duration,
+
+    listener: mpsc::Receiver<Message>,
+    sender: mpsc::Sender<Message>,
+
+    /// Reactor will dispatch messages to
+    outbound: mpsc::Sender<Message>,
+}
+
+impl TokioReactor {
+    /// Await for the next message sent from a reactor
+    pub async fn next_async(&mut self) -> Option<Message> {
+        self.listener.recv().await
+    }
+
+    /// Send a notification to the reactor
+    pub async fn notify(&mut self, notification: Notification) -> Result<(), Error> {
+        let notification = Message::Notification(notification);
+
+        self.sender
+            .send_timeout(notification, self.timeout)
+            .await
+            .map_err(|_| Error::ResourceNotAvailable)
+    }
+
+    /// Send a request to the reactor
+    pub async fn request(&mut self, request: Request) -> Result<Response, Error> {
+        let id = request.id();
+        let request = Message::Request(request);
+
+        let start = SystemTime::now();
```

Comment:
Would it make more sense to use [time::Instant](https://doc.rust-lang.org/std/time/struct.Instant.html) instead to avoid error cases of time potentially going backwards?

Reply:
These time providers libraries for rust are very flawed. Will test with time provider, but I remember there was an issue (don't remember now tho; will refresh the memory and update here)

Reply:
`time::Instant` uses same underlying OS specific time primitives as `SystemTime`.

Reply:
From the docs of [SystemTime::now().elapsed()](https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.elapsed):

> This function may fail as the underlying system clock is susceptible to drift and updates (e.g., the system clock could go backwards), so this function might not always succeed. If successful, [Ok](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Ok)([Duration](https://doc.rust-lang.org/std/time/struct.Duration.html)) is returned where the duration represents the amount of time elapsed from this time measurement to the current time.
> 
> To measure elapsed time reliably, use [Instant](https://doc.rust-lang.org/std/time/struct.Instant.html) instead.

Reply:
True, there is no point in assuming that the clock will be sync. Updated